### PR TITLE
Log-Re Pressure Scaling: Re-normalize pressure loss for OOD-Re generalization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Reynolds number in our dataset spans ~2 orders of magnitude. Pressure scales with dynamic pressure q = 0.5 * rho * V^2 ∝ Re^2 at fixed geometry (incompressible flow, unit chord). The current asinh pressure transform (merged, PR #2054) compresses extreme values but does **not** account for Re-dependent pressure scaling.

**The bet:** Our residual prediction head predicts p_node - p_freestream. At high Re, these deviations are LARGER in absolute terms, creating an implicit train-test distribution shift that specifically hurts p_re (OOD Reynolds number). If we normalize pressure loss contributions by log(Re), the target distribution becomes more Re-invariant and p_re generalization should improve.

This is a **loss normalization change only** — no architecture changes, no new input features. Concretely: the loss for each sample in the batch is divided by a mild log-Re scale factor, so high-Re samples don't dominate gradients over low-Re samples in a Re-unfair way. At inference, the model still outputs physical pressure units (inverse transform applied).

**Why log(Re) not Re or Re^2:** log(Re) is mild — it only adjusts by a factor of ~2.3× across the full Re range (10^3 to 10^5), whereas Re^2 would create a 10,000× range that destabilizes training. Log-scaling mirrors the fact that turbulence intensity scales with log(Re) in the boundary layer.

**Why this wasn't tried before:** The asinh transform already handles absolute magnitude compression. This is a complementary Re-dependent normalization addressing a different issue: not "large pressures are hard" (asinh fixes that) but "high-Re pressure fields have systematically larger magnitudes" (logRe fixes that).

**Literature:**
- Schlichting "Boundary Layer Theory" — Cp = 2(p - p_inf)/(rho U^2), all dimensionless pressure scalings
- Kashefi & Mukerji "Point-cloud deep learning of porous media for permeability prediction" (arXiv:2104.11029) — target normalization by physical scale significantly improves Re-range generalization

### Step 1: Read how Re enters the current loss

Read `cfd_tandemfoil/train.py` and find:
1. Where `re` (Reynolds number) is stored in the batch — likely in `batch.cond` or similar
2. Where the pressure MAE loss is computed for surface nodes
3. Where `asinh_pressure` transformation is applied to targets

### Step 2: Implement `--logre_pressure_scale` flag

Add a new flag `--logre_pressure_scale` (store_true). When enabled:

```python
import math

# In the loss computation, after computing pressure MAE on surface nodes:
# re is the per-sample Reynolds number, shape [B] or scalar per sample

# Compute per-sample log-Re scale factor:
# Normalize so that the scale factor is 1.0 at Re=1e5 (typical mid-range value)
log_re_scale = torch.log(re + 1.0) / math.log(1e5 + 1.0)  # shape [B], ~1.0 at Re=1e5

# Apply as per-sample weight to pressure loss (surface nodes only, pressure channel only):
# pressure_loss_per_sample has shape [B] (mean over nodes for each sample)
pressure_loss_scaled = pressure_loss_per_sample / log_re_scale  # divide by scale

# Then take mean over batch:
pressure_loss = pressure_loss_scaled.mean()
```

**Key details:**
- Apply scaling ONLY to the pressure channel in the surface node loss
- Do NOT apply to velocity field losses — pressure-only Re normalization
- Division by `log_re_scale` de-emphasizes high-Re samples (which have naturally larger pressure deviations) relative to low-Re samples
- At inference/validation: the model outputs physical pressure units directly — this is a **loss-side normalization only**, NOT a target transformation. No inverse scaling needed at eval time.
- If Re is not directly available in per-sample form, extract from `batch.cond` (the conditioning vector that also carries AoA, Umag, gap, stagger)

### Step 3: Run training

```bash
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/logre-pressure-scaling-s42" \
  --wandb_group "logre-pressure-scaling" \
  --logre_pressure_scale \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

Seed 73: same with `--seed 73` and `--wandb_name "edward/logre-pressure-scaling-s73"`.

### Step 4: If logre_pressure_scale helps p_re

If p_re improves by ≥1%, add a note in your results and I'll decide whether to sweep the normalization reference (try Re=1e4 vs Re=1e5 as the unit-scale anchor).

## Baseline

Current best metrics (PR #2251, cosine T_max=150, 2-seed avg):

| Metric | Baseline | Target |
|--------|---------|--------|
| p_in   | 11.891  | < 11.89 |
| p_oodc | 7.561   | < 7.56 |
| p_tan  | 28.118  | < 28.12 |
| p_re   | 6.364   | < 6.36 (**primary target**) |

⚠️ p_re = 6.364 is a regression vs PR #2213's 6.300. Primary goal of this experiment is to recover that regression and improve further.

Baseline W&B runs: 7jix2jkg (seed 42), epkfhxfl (seed 73).

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent <name> --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Log-Re Pressure Scaling (normalize surface pressure loss by log(Re))

| Metric | Baseline (2-seed avg) | LogRe (2-seed avg) | Delta |
|--------|----------------------|-------------------|-------|
| **p_in** | **11.891** | **12.059** | **+0.168 (worse)** |
| **p_oodc** | **7.561** | **7.662** | **+0.101 (worse)** |
| **p_tan** | **28.118** | **28.700** | **+0.582 (worse)** |
| **p_re** | **6.364** | **6.435** | **+0.071 (worse)** |

### Per-seed breakdown (Surface Pressure MAE)
| Split | s42 | s73 |
|-------|-----|-----|
| val_in_dist | 12.215 | 11.902 |
| val_ood_cond | 7.592 | 7.732 |
| val_tandem_transfer | 28.576 | 28.823 |
| val_ood_re | 6.489 | 6.381 |

### Details
- **W&B runs:** dmmemfb6 (s42), o4cnahak (s73)
- **W&B group:** logre-pressure-scaling
- **Epochs:** 149 (s42), 148 (s73)
- **Peak VRAM:** ~47 GB
- **Log-Re scale stats (train-time):** mean=1.262, std=0.014 — very small variance across batch
- **Command:**
```bash
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/logre-pressure-scaling-s42" \
  --wandb_group "logre-pressure-scaling" \
  --logre_pressure_scale \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### What happened

**Negative result.** Log-Re pressure scaling degraded all 4 metrics. The primary target p_re got 1.1% worse (6.364 → 6.435), opposite to the hypothesis.

**Analysis:**

1. **Near-zero variance in log-Re scale.** The logre_scale stats (mean=1.262, std=0.014) reveal the core problem: the scale factor barely varies across samples in the batch (std/mean ≈ 1%). This is because within a training batch, most samples come from a narrow Re range after the domain-balanced sampling. The "normalization" effectively just applies a constant ~1.26× scaling to all pressure losses — which shifts the vol/surf balance without providing any Re-adaptive benefit.

2. **Competing with PCGrad 3-way.** The 3-way PCGrad already treats extreme-Re samples as a separate loss group. Adding log-Re pressure scaling on top creates a counter-force: PCGrad tries to balance gradients from extreme-Re samples, while log-Re scaling *reduces* the loss weight of high-Re samples, reducing their gradient signal — conflicting objectives.

3. **asinh transform + residual prediction already Re-normalizes.** The asinh transform on physical pressure + subtracting the freestream already substantially reduces the absolute pressure scale variation. After these transforms, the pressure target distribution is likely already more Re-invariant than the raw pressure field, making the log-Re correction less (or counter-) productive.

4. **Division → de-emphasis of hard cases.** Dividing high-Re sample losses by a larger log-Re value means high-Re samples (the OOD-Re test cases) get systematically down-weighted during training. This reduces the model's exposure to high-Re patterns — exactly the wrong direction for improving p_re.

**The core issue:** This normalization addresses the wrong direction. If high-Re samples have larger absolute pressure errors, the right fix is to *increase* their relative weight (up-weighting OOD cases), not decrease it. The current implementation effectively does the opposite.

### Suggested follow-ups
- **Inverse log-Re scaling:** Try `surf_per_sample * log_re_scale` (multiplication instead of division) to up-weight high-Re samples — this is the opposite direction but might actually help p_re.
- **Re-stratified sampling:** Use a weighted sampler that ensures high-Re and low-Re samples are equally represented in each batch (PR #2? from nezuko).
- **Direct domain loss reweighting:** Separately track high-Re sample loss and increase its weight relative to low-Re, similar to the existing tandem_boost mechanism.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2296 because ADVISOR: Closing — all metrics regressed: p_in +2.7%, p_oodc +0.2%, p_tan +3.0%, p_re +0.2% vs current baseline (PR #2290). Excellent analysis from the student: the logRe scale has near-zero variance (std/mean ≈ 1%) across batches, making it effectively a constant multiplier. It also conflicts with PCGrad 3-way which already handles Re-based gradient balancing. The asinh transform + residual prediction + Cp normalization already provide adequate Re-normalization. Log-Re pressure scaling added to DO NOT REVISIT. W&B: dmmemfb6 (s42), o4cnahak (s73)..
